samples/nrf9160/nrf_cloud_rest_agps/src/main.c

```diff
@@ -0,0 +1,604 @@
+		.buf_sz = 0,
+		.buf = NULL
+	};
+	int err;
+
+retry:
+	err = init();
+	if (err) {
+		LOG_ERR("Initialization failed");
+		goto cleanup;
+	}
+
+	err = connect_to_network();
+	if (err) {
+		goto cleanup;
+	}
+
+	err = get_modem_info(&modem_info);
+	if (err) {
+		goto cleanup;
+	}
+	current_cell.mcc = modem_info.network.mcc.value;
+	current_cell.mnc = modem_info.network.mnc.value;
+
+	agps_print_enable(false);
+
+	err = gps_start(gps_dev, &gps_cfg);
+	if (!err) {
+		LOG_INF("GPS STARTED!");
+	}
+
+	(void)do_jitp();
+
+agps_wait:
+	LOG_INF("Waiting for AGPS request event..");
+	k_sem_take(&do_agps_request_sem, K_FOREVER);
+
+	err = modem_jwt_generate(&jwt);
+	if (err) {
+		LOG_ERR("Failed to generate JWT, err %d", err);
+		goto cleanup;
+	}
+	LOG_DBG("JWT:\n%s", log_strdup(jwt.jwt_buf));
+	rest_ctx.auth = jwt.jwt_buf;
+
+	LOG_INF("\n********************* A-GPS API *********************");
+	/* When requesting a potentially large result, set fragment size to a
+	 * small value (1) and do not provide a result buffer.
+	 * This will allow you to obtain the necessary size while limiting
+	 * data transfer.
+	 */
+	net_info.current_cell = current_cell; /* update with latest cell tower info */
+	rest_ctx.fragment_size = 1;
+	agps_sz = nrf_cloud_rest_agps_data_get(&rest_ctx, &agps_req, NULL);
+	if (agps_sz < 0) {
+		LOG_ERR("Failed to get A-GPS data: %d", agps_sz);
+		goto agps_wait; /* one might want to retry the download instead */
+	} else if (agps_sz == 0) {
+		LOG_WRN("A-GPS data size is zero, skipping");
+		goto agps_wait; /* one might want to retry the download instead */
+	}
+
+	LOG_INF("Additional buffer required to download A-GPS data of %d bytes",
+		agps_sz);
+
+	agps_result.buf_sz = (uint32_t)agps_sz;
+	agps_result.buf = k_calloc(agps_result.buf_sz, 1);
+	if (!agps_result.buf) {
+		LOG_ERR("Failed to allocate %u bytes for A-GPS buffer", agps_result.buf_sz);
+		goto cleanup;
+	}
+	/* Use the default configured fragment size */
+	rest_ctx.fragment_size = 0;
+	err = nrf_cloud_rest_agps_data_get(&rest_ctx, &agps_req, &agps_result);
+	if (err) {
+		LOG_ERR("Failed to get A-GPS data: %d", err);
+	} else {
+		/* Send the A-GPS result buffer to the modem using
+		 * gps_process_agps_data() or nrf_cloud_agps_process()
+		 */
+		LOG_INF("Successfully received A-GPS data");
+		err = agps_cloud_data_process(agps_result.buf, agps_result.agps_sz);
+		if (err) {
+			LOG_ERR("Failed to process A-GPS data: %d", err);
+		}
+	}
+
+	k_free(agps_result.buf);
+	agps_result.buf = NULL;
+	(void)nrf_cloud_rest_disconnect(&rest_ctx);
+	modem_jwt_free(jwt.jwt_buf);
+	jwt.jwt_buf = NULL;
+
+	goto agps_wait;
+
+cleanup:
+	(void)nrf_cloud_rest_disconnect(&rest_ctx);
+	modem_jwt_free(jwt.jwt_buf);
+	jwt.jwt_buf = NULL;
+
+	if (err) {
+		LOG_INF("Rebooting in 30s..");
+		k_sleep(K_SECONDS(30));
+	} else {
+		LOG_INF("Retrying in %d minute(s)..",
+			CONFIG_REST_RETRY_WAIT_TIME_MIN);
+		k_sleep(K_MINUTES(CONFIG_REST_RETRY_WAIT_TIME_MIN));
+		goto retry;
```

Comment:
does retry work as expected?  it calls init again and also jitp.

Reply:
I'm reworking it.  This was a last minute change last night.